windows/lcow_runtime.go

```diff
@@ -0,0 +1,400 @@
+	"github.com/containerd/containerd/windows/hcsshimopts"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+const (
+	lcowRuntimeName = "lcow"
+)
+
+var (
+	lcowPluginID = fmt.Sprintf("%s.%s", plugin.RuntimePlugin, lcowRuntimeName)
+)
+
+var _ = (runtime.Runtime)(&LCOWRuntime{})
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		ID:   lcowRuntimeName,
+		Type: plugin.RuntimePlugin,
+		Init: NewLCOWPlugin,
+		Requires: []plugin.PluginType{
+			plugin.MetadataPlugin,
+		},
+	})
+}
+
+func NewLCOWPlugin(ic *plugin.InitContext) (interface{}, error) {
+	if err := os.MkdirAll(ic.Root, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create state directory at %s", ic.Root)
+	}
+
+	m, err := ic.Get(plugin.MetadataPlugin)
+	if err != nil {
+		return nil, err
+	}
+
+	cd := filepath.Join(ic.Root, cacheDirectory)
+	sd := filepath.Join(ic.Root, scratchDirectory)
+
+	// Make sure the scratch directory is created under dataRoot
+	if err := os.MkdirAll(sd, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create scratch directory at %s", ic.Root)
+	}
+
+	// Make sure the cache directory is created under dataRoot
+	if err := os.MkdirAll(cd, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create cache directory at %s", ic.Root)
+	}
+
+	r := &LCOWRuntime{
+		root:    ic.Root,
+		pidPool: newPidPool(),
+
+		events:    make(chan interface{}, 4096),
+		publisher: ic.Events,
+		// TODO(mlaventure): lcow needs a stat monitor
+		monitor: nil,
+		tasks:   runtime.NewTaskList(),
+		db:      m.(*bolt.DB),
+
+		cachedSandboxFile: filepath.Join(cd, sandboxFilename),
+		cachedScratchFile: filepath.Join(cd, scratchFilename),
+		cache:             make(map[string]*cacheItem),
+		serviceVms:        make(map[string]*serviceVMItem),
+		globalMode:        false,
+	}
+
+	// Load our existing containers and kill/delete them. We don't support
+	// reattaching to them
+	r.cleanup(ic.Context)
+
+	return r, nil
+}
+
+type LCOWRuntime struct {
+	sync.Mutex
+
+	root    string
+	pidPool *pidPool
+
+	publisher events.Publisher
+	events    chan interface{}
+
+	monitor runtime.TaskMonitor
+	tasks   *runtime.TaskList
+	db      *bolt.DB
+
+	cachedSandboxFile  string                    // Location of the local default-sized cached sandbox.
+	cachedSandboxMutex sync.Mutex                // Protects race conditions from multiple threads creating the cached sandbox.
+	cachedScratchFile  string                    // Location of the local cached empty scratch space.
+	cachedScratchMutex sync.Mutex                // Protects race conditions from multiple threads creating the cached scratch.
+	serviceVmsMutex    sync.Mutex                // Protects add/updates/delete to the serviceVMs map.
+	serviceVms         map[string]*serviceVMItem // Map of the configs representing the service VM(s) we are running.
+	globalMode         bool                      // Indicates if running in an unsafe/global service VM mode.
```

Comment:
Global mode is not currently used in containerd, but will be needed in the transition to Docker. It uses a single global Service VM instead of creating a new one for each container. This is faster, but is not usable in multi-tenant environments as the shared service VM could potentially allow a malicious container to affect other containers on the system.

Reply:
Why will it be needed in the transition to Docker?

Reply:
The performance penalty of creating a new utility VM for each container is not worth the benefits to many users (such as on a private dev machine). Allowing an override to run in the faster mode is preferred in many cases, but the default should be the better security per-instance mode. Moby currently supports both.